presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -133,115 +200,440 @@ public boolean isFinished()
+                    .map(InternalAccumulatorFactory::isDistinct)
+                    .collect(toImmutableList());
+            this.intermediateTypes = accumulatorFactories.stream()
+                    .map(InternalAccumulatorFactory::getIntermediateType)
+                    .collect(toImmutableList());
+
+            this.inputs = ImmutableList.copyOf(requireNonNull(inputs, "inputs is null"));
+            this.inputSingleNullBlocks = accumulatorFactories.stream()
+                    .map(accumulatorFactory -> accumulatorFactory.getIntermediateType().createBlockBuilder(null, 1)
+                            .appendNull()
+                            .build())
+                    .collect(toImmutableList());
+        }
+
+        private int getOutputChannelCount()
+        {
+            return 1 + intermediateTypes.size() + inputs.size();
+        }
+
+        public boolean hasDistinct()
+        {
+            return intermediateIsDistinctCount > 0;
+        }
+
+        public boolean hasNonDistinct()
+        {
+            return intermediateIsDistinctCount < intermediateIsDistinct.size();
         }
 
-        public void processPage(Page page)
+        private int getIntermediateChannel(int accumulatorIndex)
         {
-            if (step.isInputRaw()) {
-                aggregation.addInput(page);
+            checkElementIndex(accumulatorIndex, intermediateTypes.size());
+            return 1 + accumulatorIndex;
+        }
+
+        private Optional<Page> buildPartialDistinctRows(Page page, List<Optional<Block>> partialDistinctMasks)
+        {
+            checkArgument(partialDistinctMasks.size() == intermediateTypes.size());
+            verify(partialDistinctMasks.stream().filter(Optional::isPresent).allMatch(mask -> mask.get().getPositionCount() == page.getPositionCount()));
+
+            // check if any positions are distinct, this shouldn't happen since functions shouldn't return an empty mask
+            SelectedPositions allDistinctPositions = combineDistinctMasks(partialDistinctMasks);
+            if (allDistinctPositions.isEmpty()) {
+                return Optional.empty();
             }
-            else {
-                aggregation.addIntermediate(page.getBlock(intermediateChannel));
+
+            List<Block> blocks = new ArrayList<>(getOutputChannelCount());
+
+            // add row type
+            blocks.add(RowType.DISTINCT_PARTIAL.createRowTypeBlock(page.getPositionCount()));
+
+            // add intermediates
+            for (int intermediateIndex = 0; intermediateIndex < partialDistinctMasks.size(); intermediateIndex++) {
+                Optional<Block> partialDistinctMask = partialDistinctMasks.get(intermediateIndex);
+                if (partialDistinctMask.isPresent()) {
+                    blocks.add(partialDistinctMask.get());
+                }
+                else if (intermediateIsDistinct.get(intermediateIndex)) {
```

Comment:
I might be missing something, but it seems to me that `intermediateIndex` may not be aligned with `intermediateIsDistinct`